src/GraphQL.Tests/Utilities/FederatedSchemaBuilderTests.cs

```diff
@@ -78,7 +78,7 @@ public void entity_query()
                 }";
 
             var variables = @"{ ""_representations"": [{ ""__typename"": ""User"", ""id"": ""123"" }] }";
-            var expected = @"{ ""_entities"": [{ ""__typename"": ""User"", ""id"" : ""123"", ""username"": ""Quinn"" }] }";
```

Comment:
This is required for federation, so we can't remove this otherwise federation would quit working.

Reply:
This is the code that modifies the selection to add `__typename` if it does not exist.

https://github.com/graphql-dotnet/graphql-dotnet/blob/f6ad2eacd9e51b08fad6805e9226ec7326e1b5d7/src/GraphQL/Utilities/Federation/FederatedSchemaBuilder.cs#L72

Reply:
Thanks would add this! I was tried to find explanation why would it is needed( to return a field nobody asking about) in docs for federation, but with no luck whatsoever...
The original query doesnt asks for it https://github.com/graphql-dotnet/graphql-dotnet/blob/acda5595ff8046b13f29323a3c1bcf37d8da8ddb/src/GraphQL.Tests/Utilities/FederatedSchemaBuilderTests.cs#L211-L219

Reply:
https://www.apollographql.com/docs/federation/federation-spec/
Federation spec doesnt impose this on entities, it only requires it for a representations. 
> Any representation without a __typename: String field is invalid.

Reply:
found somebody's sandbox with nodejs Apollo and it does exactly the same - if nobody asks for typename it doesn't returns it
![image](https://user-images.githubusercontent.com/10799080/99722880-687d5b80-2ac2-11eb-9c47-2257173f49b9.png)
Going to ask folks at Apollo to verify if **returning** `__typename` is a must, because what the point of implementing something that neither standard nor actual nodejs implementation has

Reply:
I've never used federation, but I would like to comment that Apollo GraphQL always adds `__typename` requests on each field type so that it can be used to determine proper caching rules.  So the typename field cannot be stripped off of incoming requests.

Reply:
I really don't know if that applies here; just trying to help.

Reply:
@Shane32 it doesnt get stripped when it got asked to return it
see here
https://github.com/graphql-dotnet/graphql-dotnet/blob/acda5595ff8046b13f29323a3c1bcf37d8da8ddb/src/GraphQL.Tests/Utilities/FederatedSchemaBuilderTests.cs#L128
its like whole point of GraphQL - not return what not being asked for

Reply:
sounds good. (but I really don't know anything about federation.)

Reply:
Thanks for looking more closely at this!

> its like whole point of GraphQL - not return what not being asked for

Agreed.  That has me wondering ... I swear that was required for me to get it working ... 🤔 

Perhaps it is this part of the `_Entity` union, which relies upon the `__typename`.

https://github.com/graphql-dotnet/graphql-dotnet/blob/f6ad2eacd9e51b08fad6805e9226ec7326e1b5d7/src/GraphQL/Utilities/Federation/FederatedSchemaBuilder.cs#L159-L171

I think that would get hit when the type of object returned in the resolver is different than the entity name, so that is probably a work-around.

Maybe we should add a validation rule that verifies that `__typename` is added to the representations.

Reply:
> I think that would get hit when the type of object returned in the resolver is different than the entity name, so that is probably a work-around.

I can add an unit test for this and see how it work/isn't work, because current test is more of positive test and in this case we need a negative one to be sure that it works as intended.

> Maybe we should add a validation rule that verifies that __typename is added to the representations
Should it be part of this PR?

Side note - this specific test that we talking about was the only one that wast actually working out of the box after I fixed all silly mistakes in my implementation. The issue was that type choice has to be delayed because this api returns collection of dataloader objects which type cant be figured out right away.
```cs
var gqlToNode = new Dictionary<IObjectGraphType, CompiledNode>();
foreach (var possilbe in abstractType.PossibleTypes)
{
	var fields = CollectFieldsRecursive(schema, document, variables, field.SelectionSet, possilbe);
	var rootNode = new CompiledNode(possilbe, fields);
	gqlToNode.Add(possilbe, rootNode);
}
var defaultType = new CompiledNode(abstractType, new Dictionary<string, CompiledField>());

return (value, isResolved) =>
{
	var objType = abstractType.GetObjectType(value, schema);
	return objType is { } && gqlToNode.TryGetValue(objType, out var foundType)
		? foundType
		: defaultType;
};
```

As you can see this implementation defaults to type that has no fields if type itself is a different one